doc/git-hub.1
'\" t
.\"     Title: git-hub
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 04/24/2014
.\"    Manual: Git Manual
.\"    Source: Git 1.7.9.5
.\"  Language: English
.\"
.TH "GIT\-HUB" "1" "04/24/2014" "Git 1\&.7\&.9\&.5" "Git Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
git-hub \- The Git command for GitHub \- image:https://travis\-ci\&.org/ingydotnet/git\-hub\&.png[link="https://travis\-ci\&.org/ingydotnet/git\-hub"]
.SH "SYNOPSIS"
.sp
Here are just a few of the things you can do on GitHub with a simple command:
.sp
.nf
git hub                             # Show the `git hub` cheat sheet
git hub help                        # Show this awesome documentation
git hub setup                       # Config dialog\&. Do this first!
.fi
.sp
.nf
git hub user <user>                 # Get info about a user
git hub repos <owner>               # List recent repositories
git hub repo\-new <repo>             # Create a new repository
git hub fork <owner>/<repo>         # Fork a repository
git hub clone <owner>/<repo>        # Clone a GitHub repository
                                    # Set/Change your repo description
git hub repo\-edit <repo> description \*(AqJust another GitHub repo\*(Aq
git hub repo\-delete <owner>/<repo>  # Delete a repository
git hub followers <user>            # List a user\*(Aqs followers
git hub star ingydotnet/git\-hub     # Star this repository
git hub follow ingydotnet           # Follow this author
git hub trust ingydotnet            # Add collaborators to repo
                                    # Stalk all your followers
git hub followers \-ra | git hub user \-
.fi
.SH "DESCRIPTION"
.sp
The \fBhub\fR subcommand for git, allows you to perform many of the operations made available by GitHub\(cqs v3 REST API, from the git commandline command\&.
.sp
You can fork, create, delete and modify repositories\&. You can get information about users, repositories and issues\&. You can \fBstar\fR, \fBwatch\fR and \fBfollow\fR things, and find out who else is doing the same\&. The API is quite extensive\&. With this command you can do many of your day to day GitHub actions without needing a web browser\&.
.sp
You can also chain commands together using the output of one as the input of another\&. For example you could use this technique to clone all the repos of a GitHub user or organization, with one command\&. See PIPE AND EACH below\&.
.SH "INSTALLATION"
.sp
Get the source code from GitHub:
.sp
.if n \{\
.RS 4
.\}
.nf
git clone git@github\&.com:ingydotnet/git\-hub
.fi
.if n \{\
.RE
.\}
.sp
Then simply add the lib and man directories to your PATH and MANPATH:
.sp
.if n \{\
.RS 4
.\}
.nf
export PATH=/path/to/git\-hub/lib:$PATH
export MANPATH=/path/to/git\-hub/man:$MANPATH
.fi
.if n \{\
.RE
.\}
.sp
To do a proper system install, which puts git\-hub next to your other git commands, do this instead:
.sp
.if n \{\
.RS 4
.\}
.nf
make install        # Possibly with \*(Aqsudo\*(Aq
.fi
.if n \{\
.RE
.\}
.SH "UPGRADING"
.sp
If you used the PATH method of installation, just run this to upgrade git\-hub:
.sp
.if n \{\
.RS 4
.\}
.nf
git hub upgrade
.fi
.if n \{\
.RE
.\}
.sp
If you used make install method, then:
.sp
.if n \{\
.RS 4
.\}
.nf
cd /path/to/git\-hub
git pull
make install
.fi
.if n \{\
.RE
.\}
.SH "OPTIONS"
.sp
Commandline options start with a dash and can appear in any order in the command\&.
.PP
\-c <number>, \-\-count=<number>
.RS 4
This option is for commands that return a list of things\&. Specify the number of items you want the command to return\&. Each list command has a reasonable default value\&.
.RE
.PP
\-a, \-\-all
.RS 4
Show all elements for list commands\&. Also applies to a few other commands as documented\&.
.RE
.PP
\-r, \-\-raw
.RS 4
For commands the produce lists, just list the raw data\&. For commands that produce reports, show the raw storage keys, rather than the human friendly ones\&. You will need the raw keys to edit/change values\&.
.RE
.PP
\-j, \-\-json
.RS 4
For some commands, you can get the raw data in JSON\&.
.RE
.PP
\-
.RS 4
This option, just a single
\fI\-\fR, is small but very powerful\&. It means to read the next line from stdin, and put the content into the command at that point\&. See PIPE and EACH below\&.
.RE
.PP
=
.RS 4
Like the \*(Aq\-\*(Aq\ \&command above, but reads each
\fBword\fR
from stdin\&.
.RE
.PP
\-h
.RS 4
Show the brief command usage\&.
.RE
.PP
\-\-help
.RS 4
Show the full documentation\&.
.RE
.PP
\-q, \-\-quiet
.RS 4
Show minimal output\&.
.RE
.PP
\-v, \-\-verbose
.RS 4
Show detailed output\&. Notably, show the raw command that is used for the GitHub API calls\&.
.RE
.PP
\-\-token=<api\-token>
.RS 4
Set or override the api\-token required by the command\&. This value can also come from the
GIT_HUB_API_TOKEN
environment variable\&. Usually this value comes from your
~/\&.githubconfig
file\&.
.RE
.PP
\-d, \-\-dryrun
.RS 4
Don\(cqt actually make the API call\&. This option turns on verbose mode\&.
.RE
.PP
\-T
.RS 4
Show the API token in the verbose output\&. Hidden by default\&. This option turns on verbose mode\&.
.RE
.PP
\-A, \-\-use\-auth
.RS 4
Use authentication whether or not it is needed for a certain command\&. This is useful if you have exceeded your hourly GitHub API rate limit, which is 60 calls unauthenticated, and 5000 calls authenticated\&.
.RE
.PP
\-C, \-\-no\-cache
.RS 4
There are currently known bugs in GitHub\(cqs cache headers\&. If the content seems stale, use this option\&.
.RE
.SH "DEBUG OPTIONS"
.sp
There are a number of debug options that will give you lots of extra information (albeit in a less pretty format)\&.
.PP
\-O
.RS 4
Show the API reponse output\&. Always in JSON format\&. JSON is pretty\-printed\&.
.RE
.PP
\-H
.RS 4
Show the API response headers\&.
.RE
.PP
\-J
.RS 4
Show the API response JSON, in decoded form\&.
.RE
.PP
\-R
.RS 4
Repeat the last command, using the previous API server response\&. Useful for development and debugging\&. (Beware, only caches the most recent API call, so not really good for commands that make multiple calls)\&.
.RE
.PP
\-x
.RS 4
Turn on Bash
\-x
debugging\&. This will show every bash command executed in the program\&.
.RE
.SH "ARGUMENTS"
.sp
The git hub command gets its input values from these sources (in this order):
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Commandline
\fBarguments\fR
(see specific command, below)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Environment variables of the form
GIT_HUB_VARIABLE_NAME\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The
\&./\&.git/config
file\&. (The repo you are issuing commands from)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The
~/\&.githubconfig
file\&.
.RE
.sp
Most of the commands described below have arguments\&. This section defines each of the argument types\&. Note that argument values that are strings containing whitespace need to be quoted\&.
.sp
Argument variable names are always indicated by angle brackets, like: <variable>\&. Arguments listed in square brackets are optional and arguments followed by \&.\&.\&. indicate that more than one is allowed\&.
.PP
<user>
.RS 4
A GitHub user name\&. Sometimes an organization name can be used instead\&. If this argument is optional, it defaults to the GitHub owner of the repo you are currently in\&. If you are not in a GitHub repo, it defaults to your login\&.
\fBOverride\fR:
GIT_HUB_USER_NAME
environment variable\&.
.RE
.PP
<owner>
.RS 4
A GitHub user or organization that owns the repository being specified\&. This value defaults much like the
<user>
argument (above)\&.
\fBOverride\fR:
GIT_HUB_OWNER_NAME
environment variable\&.
.RE
.PP
<org>
.RS 4
A GitHub organization name\&.
\fBOverride\fR:
GIT_HUB_ORG_NAME
environment variable\&.
.RE
.PP
<repo>
.RS 4
A GitHub repository name\&. If this argument is optional, it defaults to the GitHub repo name of the repo you are currently in\&.
\fBOverride\fR:
GIT_HUB_REPO_NAME
environment variable\&.
.RE
.PP
<owner>/<repo>
.RS 4
Many commands need both an owner and repo separated by a slash\&. If the owner is you (your GitHub login) you can omit it (but you still need the leading slash)\&. Like
/myrepo, instead of
me/myrepo\&. If this argument is optional, it defaults the the owner and repo of the GitHub repo you are in\&.
.RE
.PP
<key\-value\-pair>
.RS 4
Two strings separated by whitespace\&. Values with embedded whitespace should be quoted\&. The documentation of each command that needs these pairs, will list the valid keys\&.
.RE
.PP
<api\-token\-id>
.RS 4
The integer number
\fBid\fR
of a token (not the 40 hex\-character value)\&.
.RE
.PP
<scope\-name>
.RS 4
One of the GitHub scopes that you can apply to a token\&. The
git hub scope\-list
command will tell you all of these, and what they mean\&.
.RE
.SH "COMMANDS"
.sp
These are the commands you can use to perform most of your daily GitHub interactions from the command line\&.
.PP
setup
.RS 4
Before you can use the commands described below, you need to perform some setup/configuration steps\&.
.sp
.if n \{\
.RS 4
.\}
.nf
This "wizard" style dialog, will walk you through the configuration
process quickly and painlessly, with lots of explanation\&. You should run
this command right away\&. You can also re\-run it, and it will allow you to
change your config, while defaulting to your existing settings\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
If you would rather do the steps by hand, see CONFIGURATION COMMANDS
below\&.
.fi
.if n \{\
.RE
.\}
.RE
.PP
user [<user>]
.RS 4
Show basic information about a specific user\&. User defaults to the owner of the current repo, or your login if you are not inside a repo directory\&. The
\-\-raw
and
\-\-json
options show the data in different formats than normal output\&.
.RE
.PP
user\-get <user> <data\-key>
.RS 4
Get a specific data value for a particular user\&.
.RE
.PP
user\-edit <key\-value\-pair>\&...
.RS 4
Set specific fields of your user info to new values\&. You list the parameters as key/value pairs\&.
.sp
.if n \{\
.RS 4
.\}
.nf
You can edit the following user keys: `name`, `email`, `blog`,
`location`, `company`, `bio`\&.
.fi
.if n \{\
.RE
.\}
.RE
.PP
orgs [<user>]
.RS 4
List the organizations that a user is a member of\&.
.RE
.PP
org <org>
.RS 4
Show basic information about a GitHub organization\&. The
\-\-raw
and
\-\-json
options show the data in different formats than normal output\&.
.RE
.PP
org\-get <org> <data\-key>
.RS 4
Get a specific data value for a particular organization\&.
.RE
.PP
org\-edit <org> <key\-value\-pairs>\&...
.RS 4
Set specific meta\-data fields of an organization to new values\&. You list the parameters as key/value pairs\&.
.sp
.if n \{\
.RS 4
.\}
.nf
You can edit the following organization keys: `name`, `email`,
`billing_email`, `blog`, `location`, `company`\&.
.fi
.if n \{\
.RE
.\}
.RE
.PP
members <org>
.RS 4
List the members of an organization\&.
.RE
.PP
teams <org>
.RS 4
List the teams in an organization\&.
.RE
.PP
followers [<user>]
.RS 4
List the people who are followers of a user\&.
.RE
.PP
follows <user> [<target\-user>]
.RS 4
Check if <user> follows <target\-user>\&. The default target user is you\&.
.RE
.PP
following [<user>]
.RS 4
List the people that a user is following\&.
.RE
.PP
follow <user>\&...
.RS 4
Follow one or more users\&.
.RE
.PP
unfollow <user>\&...
.RS 4
Stop following one or more users\&.
.RE
.PP
clone ([<owner>/]<repo> [<directory>])\&...
.RS 4
Clone a GitHub repo\&. Always uses
\-\-recursive
so you get submodules too\&. You can specifiy a list of repos\&. If you specify a directory for a repo, it should be an absolute path name or else begin with
\fI\&./\fR
or
\fI\&.\&./\fR
so that it can be distinguished from another repo name\&.
.RE
.PP
repos [<user>]
.RS 4
List the repos for a user or organization\&. List is returned in order of recent activity\&.
.RE
.PP
repo [<repo>]
.RS 4
Show basic information about a specific repository\&. The
\-\-raw
and
\-\-json
options show the data in different formats than normal output\&.
.RE
.PP
repo\-get <owner>/<repo> <data\-key>
.RS 4
Get a specific data value for a particular repository\&.
.RE
.PP
repo\-edit [<owner>/<repo>] <key\-value\-pair>\&...
.RS 4
Set specific meta\-data fields of a repository to new values\&. You list the parameters as key/value pairs\&.
.sp
.if n \{\
.RS 4
.\}
.nf
You can edit the following repo keys: `description`, `homepage`\&.
.fi
.if n \{\
.RE
.\}
.RE
.PP
repo\-new [<org>/]<repo>
.RS 4
Create a new GitHub repository\&.
.RE
.PP
repo\-delete <owner>/<repo>
.RS 4
Delete a GitHub repository\&.
.RE
.PP
forks [<owner>/<repo>]
.RS 4
List the forks of a repository\&.
.RE
.PP
fork <owner>/<repo> [<org>]
.RS 4
Fork a repository to your account or to an organization\&.
.RE
.PP
stars [<owner>/<repo>]
.RS 4
Show what users have starred a repository\&.
.RE
.PP
star [<owner>/<repo>]
.RS 4
Add your
\fBstar\fR
to a repository\&.
.RE
.PP
unstar [<owner>/<repo>]
.RS 4
Add your
\fBstar\fR
to a repository\&.
.RE
.PP
starred [<user>]
.RS 4
List repositories that a user has starred\&.
.RE
.PP
collabs [<owner>/<repo>]
.RS 4
List current collaborators for a repository\&.
.RE
.PP
trust [<owner>/<repo>] <user>\&...
.RS 4
Add one or more collaborators to a repository\&.
.RE
.PP
untrust [<owner>/<repo>] <user>\&...
.RS 4
Remove one or more collaborators from a repository\&.
.RE
.PP
issues [<owner>/<repo>]
.RS 4
List the issues for a repo\&.
.RE
.PP
issue [<owner>/<repo>] <issue\-id\-number>
.RS 4
Show info (including any comments) for a specific issue\&.
.RE
.PP
issue\-new [<owner>/<repo>]
.RS 4
Create a new issue for a repository\&.
.RE
.PP
issue\-comment [<owner>/<repo>] <issue\-id\-number> [<comment\-text\-string>]
.RS 4
Add a comment to an issue\&. If no comment string is given, the command will open your default editor and let you enter the text\&.
.RE
.PP
issue\-close [<owner>/<repo>] <issue\-id\-number> [<comment\-text\-string>]
.RS 4
Close an issue with an option comment\&. If you use the string
\fI\&...\fR
as your comment, your ediitor will be opened to allow you to type in your comment text\&.
.RE
.PP
keys [<user>]
.RS 4
List public SSH keys for a user\&.
.RE
.SH "CONFIGURATION COMMANDS"
.sp
These commands are for editing your git\-hub config file and managing your GitHub authentication tokens\&.
.PP
config [<config\-key> [<config\-value>]]
.RS 4
With no args, this command will print the contents of

Changes
---
- version: 0.0.1
  date:
  changes:


src/git-hub.bash
#!/bin/bash
#
# git-hub: Use the GitHub v3 API commands from git
#
# Copyright (c) 2013 Ingy döt Net

set -e

GIT_HUB_VERSION=0.0.1

OPTIONS_SPEC="\
git hub <command> <options> <arguments>

Commands:
  config
  user-info, user-update
  repo-list, repo-info, repo-create, repo-edit, repo-delete
  collab-list, collab-add, collab-remove

Aliases:
  user == user-info
  repos == repo-list
  repo == repo-info

See 'git help hub' for complete documentation and usage of each command.

Options:
--
user=       GitHub user name
repo=       GitHub repository name
token=      GitHub v3 API Authentication Token
c,count=    Number of items to process

h,help      Show this help
q,quiet     Show minimal output
v,verbose   Show verbose output
d,dryrun    Don't run the API command
T           Show API token in the verbose output
 
O           dev - Show response output
H           dev - Show reponse headers
J           dev - Show parsed JSON response
x           dev - Turn on Bash trace (set -x) output
R           dev - Repeat last command without contacting server
"

#------------------------------------------------------------------------------
main() {
	assert-env
	get-options "$@"
	setup-env

	! callable github-"$command" &&
		die "Unknown 'git hub' command: '$command'"
	github-"$command" "$@"
	[ -n "$show_headers" ] && cat $GIT_HUB_HEADER
	[ -n "$show_output" ] && cat $GIT_HUB_OUTPUT
	[ -s $GIT_HUB_OUTPUT ] && json-load-cache "$(< $GIT_HUB_OUTPUT)"
	[ -n "$show_json" ] && echo "$json_load_data_cache"

	if OK; then
		if callable "success-github-$command"; then
			"success-github-$command"
		else
			say ${message_success:-"'git hub $command' successful"}
		fi
		exit 0
	elif [ -n "$status_code" ] && callable "status-$status_code-github-$command"; then
		"status-$status_code-github-$command"
		exit 1
	elif [ -n "$(eval "echo \$message_$status_code")" ]; then
		say $(eval "echo \$message_$status_code")
		exit 1
	elif callable "failure-github-$command"; then
		"failure-github-$command"
		exit 1
	else
		say ${message_failure:="'git hub $command' failed: $status_code $ERROR"}
		exit 1
	fi
}

#------------------------------------------------------------------------------
# `git hub` command functions:
#------------------------------------------------------------------------------
github-config() {
    get-args config_key config_value
	if [ -z "$config_key" ]; then
		cat $config_file
	elif [ -z "$config_value" ]; then
		message_success=$(git config -f $config_file github.$config_key)
	else
		git config -f $config_file github.$config_key "$config_value"
		message_success="$config_key=$config_value"
	fi
	OK=$?
}

github-user-info() {
    get-args user
	require-value user-name "$user"
	api-get "/users/$user_name"
}

success-github-user-info() {
	for field in \
		login type name email blog location company \
		followers following public_repos public_gists
	do
		report-value $field
	done
}

github-user-update() {
	get-args *key_value_pairs
	api-patch "/user" "$(json-dump-object-pairs)"
}

github-repo-list() {
    get-args user
	require-value user-name "$user"
	page_size=100
	per_page=$list_count
	[ $per_page -gt $page_size ] && per_page=$page_size
	api-get "/users/$user_name/repos?sort=pushed;per_page=$per_page"
	counter=1
}

success-github-repo-list() {
	for ((ii=0; ii < $page_size; ii++)); do
		[ $counter -le $list_count ] || return 0
		name=$(json-get "/$ii/full_name")
		pushed=$(json-get "/$ii/pushed_at")
		pushed=${pushed/T*/}
		desc=$(json-get "/$ii/description")
		[ -z "$name" ] && break
		printf "%3d) (%s)  %-30s %s\n" $((counter++)) $pushed $name "$desc"
	done
	if [ $ii -ge $page_size ]; then
		get-next-page success-github-repo-list
	fi
}

github-repo-info() {
	get-args repo
	check-user-repo "$repo"
	require-value repo-name "$repo"
	require-value user-name "$user"
	api-get "/repos/$user_name/$repo_name"
}

success-github-repo-info() {
	for field in \
		full_name description homepage language \
		pushed_at \
		html_url ssh_url \
		forks watchers
	do
		report-value $field
	done
}

github-repo-create() {
	get-args repo
	require-value repo-name "$repo"
	api-post "/user/repos" $(json-dump-object 'name' $repo_name)
	message_success="Repository '$repo_name' created."
	message_422="Repository name '$repo_name' already exists."
}

github-repo-edit() {
	get-args *key_value_pairs
	require-value repo-name "$repo"
	require-value user-name "$user"
	key_value_pairs+=(name "$repo_name")
	api-patch "/repos/$user_name/$repo_name" "$(json-dump-object-pairs)"
}

github-repo-delete() {
	get-args repo
	check-user-repo "$repo"
	require-value repo-name "$repo"
	require-value user-name "$user"
	api-delete "/repos/$user_name/$repo_name"
	message_success="Repository '$repo_name' deleted"
}

github-collab-list() {
	get-args user-repo
	require-value repo-name "$repo"
	require-value user-name "$user"
	page_size=100
	per_page=$list_count
	[ $per_page -gt $page_size ] && per_page=$page_size
	api-get "/repos/$user_name/$repo_name/collaborators?per_page=$per_page"
	counter=1
}

success-github-collab-list() {
	for ((ii=0; ii < $page_size; ii++)); do
		[ $counter -le $list_count ] || return 0
		name=$(json-get "/$ii/login")
		[ -z "$name" ] && break
		printf "%3d) %s\n" $((counter++)) $name
	done
	if [ $ii -ge $page_size ]; then
		get-next-page success-github-collab-list
	fi
}

github-collab-add() {
	say "*** NOTE *** Can't get collab-add working yet. Patches welcome."
	get-args user-repo *collaborators
	require-value repo-name "$repo"
	require-value user-name "$user"
	for collab_name in ${collaborators[@]}; do
		api-put "/repos/$user_name/$repo_name/collaborators/$collab_name"
	done
}

github-collab-remove() {
	get-args user-repo *collaborators
	require-value repo-name "$repo"
	require-value user-name "$user"
	for collab_name in ${collaborators[@]}; do
		api-delete "/repos/$user_name/$repo_name/collaborators/$collab_name"
	done
}
#------------------------------------------------------------------------------
# API calling functions:
#------------------------------------------------------------------------------
api-get() { api-call GET "$1" "$2"; }
api-post() { api-call POST "$1" "$2"; }
api-put() { api-call PUT "$1" "$2"; }
api-patch() { api-call PATCH "$1" "$2"; }
api-delete() { api-call DELETE "$1" "$2"; }

# Build a command to make the HTTP call to the API server, make the call, and
# check the result.
api-call() {
	[ -n "$repeat_command" ] && api-repeat && return

	require-value api-token "$token"
	local action=$1
	local url=$2
	local data=$3
	local regex="^https?:"
	[[ "$url" =~ $regex ]] || url="$GIT_HUB_API_URI$url"

	# Build curl command in an array. This is the only way to preserve quoted
	# whitespace.
	local command=(
		curl --silent --show-error --request $action
		--user-agent git-hub-$GIT_HUB_VERSION
	)
	command+=(--header "Authentication: token $api_token")
	command+=(
		"${auth[@]}"
		$url
	)
	[ -n "$data" ] && command+=(-d "$data")
	command+=(
		--dump-header $GIT_HUB_HEADER
		--output $GIT_HUB_OUTPUT
		--stderr $GIT_HUB_ERROR
	)

	if [ $GIT_VERBOSE ]; then
		if [ -n "$show_token" ]; then
			say "${command[@]}"
		else
			say "${command[@]/$api_token/********}"
		fi
	fi

	[ -n "$dry_run" ] && exit 0

	# Run the curl command!
	"${command[@]}"
	## Use this line for debugging the real command
	# set -x; "${command[@]}"; set +x

	check-api-call-status $?
}

check-api-call-status() {
	OK=$1
	if [ ! -f $GIT_HUB_HEADER ]; then
		OK=1
		ERROR=$(head -1 $GIT_HUB_ERROR)
		return
	fi
	status_code=$(head -1 $GIT_HUB_HEADER | cut -d ' ' -f2)
	local regex='^[0-9]{3}$'
	[[ $status_code =~ $regex ]] || return
	case "$status_code" in
		200|201|204) OK=0 ;;
		*)
			OK=1
			ERROR=$(head -1 $GIT_HUB_HEADER | cut -d ' ' -f3-)
			;;
	esac
}

OK() {
	return $OK
}

api-repeat() {
	check-api-call-status 0
	true
}

#------------------------------------------------------------------------------
# Argument parsing functions:
#------------------------------------------------------------------------------
get-args() {
	local slurp_index=0
    local slurp_re='^\*'
	local user_repo_re='/'
	if [[ "$1" = "user-repo" ]]; then
		if [[ ${command_arguments[0]} =~ $user_repo_re ]]; then
			check-user-repo ${command_arguments[0]}
			unset command_arguments[0]
		fi
		shift
	fi
	for arg in "${command_arguments[@]}"; do
		if [ $slurp_index -gt 0 ]; then
			eval ${1/\*/}[$slurp_index]=\"$arg\"
			: $((slurp_index++))
		elif [ $# -gt 0 ]; then
			if [[ $1 =~ $slurp_re ]]; then
				eval ${1/\*/}[$slurp_index]=\"$arg\"
				: $((slurp_index++))
			else
				eval $1="$arg"
				shift
			fi
		else
			die "Unknown argument: $arg"
		fi
	done
}

check-user-repo() {
	[[ $1 =~ "/" ]] && user-repo ${1/\// }
	true
}

user-repo() {
	user=$1
	repo=$2
}

get-next-page() {
    local callback=$1
	regexp='Link: <(https:.+?)>; rel="next"'
	[[ "$(< $GIT_HUB_HEADER)" =~ $regexp ]] || return
	local link=${BASH_REMATCH[1]}
	api-get "$link"
	if OK; then
		json-load-cache "$(< $GIT_HUB_OUTPUT)"
		$callback
	fi
}

# Usage: require-value variable-name "possible-user-value"
#   Fetch $variable_name or die
require-value() {
	local key=$1
	local var=${key//-/_}
	fetch-value "$@"
	if [ -z "$(eval echo \$$var)" ]; then
		[ "$var" = "api_token" ] && die_need_api_token
		die "Can't find value for '$var'"
	fi
	true
}

# Usage: fetch-value variable-name "possible-user-value"
#   Sets $variable_name to the first of:
#   - possible-user-value
#   - $GIT_HUB_VARIABLE_NAME
#   - git config github.variable-name
fetch-value() {
	local key=$1
	local var=${key//-/_}
	local env=GIT_HUB_$(echo $var | tr 'a-z' 'A-Z')

	[ -n "$(eval echo \$$var)" ] && return
	eval $var="$2"
	[ -n "$(eval echo \$$var)" ] && return
	eval $var=\"\$$env\"
	[ -n "$(eval echo \$$var)" ] && return
	if [ "$var" = "repo_name" -o "$var" = "user_name" ]; then
		if [ -f ".git/config" ]; then
			url=$(git config --file=.git/config remote.origin.url)
			if [ -n "$url" ]; then
				local re1='github\.com'
				if [[ "$url" =~ $re1 ]]; then
					local re2='^.*[:/](.*)/(.*)\.git$'
					if [[ "$url" =~ $re2 ]]; then
						: ${user_name:="${BASH_REMATCH[1]}"}
						: ${repo_name:="${BASH_REMATCH[2]}"}
						return
					fi
				fi
			fi
		fi
	fi
	eval $var=$(git config --file=$config_file github.$key || echo '')
	[ -n "$(eval echo \$$var)" ] && return
	true
}

#------------------------------------------------------------------------------
# Detailed error messages:
#------------------------------------------------------------------------------
die_need_api_token() {
	cat <<eos

Can't determine your API Access Token. Usually this means you haven't set up
your ~/.githubconfig file yet.

First go to https://github.com/settings/applications and retrieve or create a
Personal API Access Token. This is a 40 digit hexadecimal character string.

Next, run these commands:

	git hub config api-token <your-personal-api-access-token>
	git hub config user-name <your-github-login-id>

Now you should be set up to run most commands. Some commands require that you
add certain 'scopes' to your token.

eos
	die
}

#------------------------------------------------------------------------------
# JSON support functions:
#------------------------------------------------------------------------------
# Format a JSON object from an input list of key/value pairs.
json-dump-object() {
	local json='{'
	while [ $# -gt 0 ]; do
		json="$json\"$1\":\"$2\""
		shift; shift
		if [ $# -gt 0 ]; then
			json="$json,"
		fi
	done
	json="$json}"
	echo $json
}

# Format a JSON object from an array.
json-dump-object-pairs() {
	local json='{'
	for ((i = 0; i < ${#key_value_pairs[@]}; i = i+2)); do
		local value=${key_value_pairs[$((i+1))]}
		json="$json\"${key_value_pairs[$i]}\":\"$value\""
		if [ $((${#key_value_pairs[@]} - $i)) -gt 2 ]; then
			json="$json,"
		fi
	done
	json="$json}"
	echo $json
}

json-load-cache() {
	json_load_data_cache=$(echo "$1" | tokenize | parse)
}

json-get() {
	local key="^$1\s"
	local data="$2"
	[ -z "$data" ] && data="$json_load_data_cache"
	value=$(echo "$data" | egrep $key | cut -f2)
	if [ "$value" = 'null' ]; then
		echo ''
	else
		echo ${value//\"/}
	fi
}

#------------------------------------------------------------------------------
# Report formatting functions:
#------------------------------------------------------------------------------
report-value() {
	local value=$(json-get "/$1")
	local label=$(eval echo \$label_$1)
	if [ -z "$label" ]; then
		label=$(echo "$1" | tr '_' ' ')
		label=${label//ssh/SSH}
		label=${label//url/URL}
		label=${label//html/HTML}
		label=$(for word in $label; do title=`echo "${word:0:1}" | tr a-z A-Z`${word:1}; echo -n "$title "; done)
	fi
	if [ -n "$label" -a -n "$value" ]; then
		printf "%-15s %s\n" "$label" "$value"
	fi
}

label_blog='Web Site'

#------------------------------------------------------------------------------
# Initial setup functions:
#------------------------------------------------------------------------------
assert-env() {
	GIT_CORE_PATH=$(git --exec-path) || exit $?
	PATH=$GIT_CORE_PATH:$PATH

	: ${GIT_HUB_API_URI:=https://api.github.com}
	: ${GIT_HUB_TMP_DIR:=/tmp}
	GIT_HUB_TMP_PREFIX=$GIT_HUB_TMP_DIR/git-hub
	GIT_HUB_INPUT=$GIT_HUB_TMP_PREFIX-in-$$
	GIT_HUB_OUTPUT=$GIT_HUB_TMP_PREFIX-out-$$
	GIT_HUB_ERROR=$GIT_HUB_TMP_PREFIX-err-$$
	GIT_HUB_HEADER=$GIT_HUB_TMP_PREFIX-head-$$
	local cmd
	for cmd in git curl tr egrep head cut; do
		[ -z "$(which $cmd)" ] &&
			echo "Required command not found: '$cmd'" && exit 1
	done
	[ -z "$HOME" ] &&
		echo "Cannot determine HOME directory" && exit 1
	config_file=$GIT_HUB_CONFIG
	[ -n "$config_file" ] || config_file=$HOME/.githubconfig
}

get-options() {
	[ $# -eq 0 ] && set -- --help
	NONGIT_OK=1 source git-sh-setup

	user=; repo=; token=; list_count=10
	GIT_QUIET=; GIT_VERBOSE=; show_token=
	user=; repo=; dry_run=;
	while [ $# -gt 0 ]; do
		local option="$1"; shift
		case "$option" in
			--user) user="$1"; shift ;;
			--repo) repo="$1"; shift ;;
			--token) token="$1"; shift ;;
			-c)	list_count=$1; shift ;;

			-h) usage ;;
			-q) GIT_QUIET="1"
				GIT_VERBOSE=
				;;
			-v) GIT_VERBOSE="1"
				GIT_QUIET=
				;;
			-d) dry_run="1" ;;
			-T) show_token="1" ;;

			--) break ;;

			# Dev options:
			-O) show_output="1" ;;
			-H) show_headers="1" ;;
			-J) show_json="1" ;;
			-x) set -x ;;
			-R) repeat_command="1" ;;

			*) die "Unexpected option: $option" ;;
		esac
	done

    command="$1"; shift
	# Some common command aliases:
	[ "$command" = "user" ] && command="user-info"
	[ "$command" = "repo" ] && command="repo-info"
	[ "$command" = "repos" ] && command="repo-list"

	# TODO find better idiom than 'eval' to global array
	for ((i = 0; i < $#; i++)); do
		eval command_arguments[$i]=\$$((i+1))
	done
}

setup-env() {
    source "$GIT_CORE_PATH/lib/json.bash"
    if [ -n "$repeat_command" ]; then
		[ -f $GIT_HUB_TMP_PREFIX-out-* ] ||
			die "No previous 'git hub' command to repeat"
		local old_output=$(echo $GIT_HUB_TMP_PREFIX-out-*)
		local pid=${old_output/$GIT_HUB_TMP_PREFIX-out-/}
		GIT_HUB_INPUT=$GIT_HUB_TMP_PREFIX-in-$pid
		GIT_HUB_OUTPUT=$GIT_HUB_TMP_PREFIX-out-$pid
		GIT_HUB_ERROR=$GIT_HUB_TMP_PREFIX-err-$pid
		GIT_HUB_HEADER=$GIT_HUB_TMP_PREFIX-head-$pid
	else
		rm -f $GIT_HUB_TMP_PREFIX-*
	fi
	[ -n "$dry_run" ] &&
		say '*** NOTE: This is a dry run only. ***'
	true
    # require_work_tree
}


#------------------------------------------------------------------------------
# General purpose functions:
#------------------------------------------------------------------------------

# Check if a function exists:
callable() {
	[ -n "$(type $1 2> /dev/null)" ]
}

#------------------------------------------------------------------------------
# Begin at the end!
#------------------------------------------------------------------------------
main "$@"

# vim: set tabstop=4 shiftwidth=4 noexpandtab:
